gui/src/ruler.cc

```diff
@@ -0,0 +1,362 @@
+#include <cmath>
+#include <iostream>
+#include "ruler.hh"
+
+////////////////////////////////////////////////////////////////////////
+// Ruler
+
+Ruler::Ptr Ruler::create(Ruler::Orientation orientation, GtkWidget *drawingArea)
+{
+    Ruler::Ptr ruler{new Ruler(orientation)};
+    if (drawingArea != nullptr)
+    {
+        ruler->setDrawingArea(drawingArea);
+    }
+    ruler->update();
+    return ruler;
+}
+
+Ruler::Ruler(Ruler::Orientation orientation)
+        : orientation{orientation}
+{
+}
+
+Ruler::~Ruler()
+{
+    unregisterDrawingArea();
+}
+
+void Ruler::setDrawingArea(GtkWidget *widget)
+{
+    unregisterDrawingArea();
+    drawingArea = widget;
+
+    width = gtk_widget_get_allocated_width(widget);
+    height = gtk_widget_get_allocated_height(widget);
+
+    // Register callbacks
+    g_signal_connect(widget, "draw", G_CALLBACK(drawCallback), this);
+    g_signal_connect(widget, "size-allocate", G_CALLBACK(sizeAllocateCallback), this);
+}
+
+void Ruler::unregisterDrawingArea()
+{
+    if (drawingArea == nullptr) { return; }
+
+    g_signal_handlers_disconnect_by_data(drawingArea, this);
+    drawingArea = nullptr;
+}
+
+void Ruler::setRange(double lower, double upper)
+{
+    lowerLimit = lower;
+    upperLimit = upper;
+
+    if (drawingArea == nullptr) { return; }
+
+    update();
+
+    // We need to manually trigger the widget to redraw
+    gtk_widget_queue_draw(drawingArea);
+}
+
+void Ruler::update()
```

Comment:
I was expecting `update()` to somehow trigger a redraw. It seems that it doesn't. That's fine, but then maybe you want to change the name to indicate what exactly it is you are updating.